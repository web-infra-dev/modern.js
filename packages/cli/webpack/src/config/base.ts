/* eslint-disable max-lines */
import path from 'path';
import {
  chalk,
  isProd,
  isDev,
  logger,
  CHAIN_ID,
  isString,
  ensureArray,
  isTypescript,
  ensureAbsolutePath,
  applyOptionsChain,
  removeLeadingSlash,
} from '@modern-js/utils';
import webpack from 'webpack';
import type { IAppContext, NormalizedConfig } from '@modern-js/core';
import WebpackChain from '@modern-js/utils/webpack-chain';
import type { Options as BabelPrestAppOptions } from '@modern-js/babel-preset-app';
import { merge as webpackMerge } from '../../compiled/webpack-merge';
import { JS_RESOLVE_EXTENSIONS } from '../utils/constants';
import { enableCssExtract } from '../utils/createCSSRule';
import { getWebpackLogging } from '../utils/getWebpackLogging';
import { ChainUtils, getWebpackUtils } from './shared';
import { applyTsLoader, applyTsCheckerPlugin } from './features/ts';
import { applyBabelLoader } from './features/babel';
import { applyModuleScopePlugin } from './features/module-scope';
import { applyMinimizer } from './features/minimizer';
import { applySvgrLoader } from './features/svgr';
import { applyAlias, applyTsConfigPathsPlugins } from './features/alias';
import { applyAssetsLoader } from './features/assets';
import { applyCSSExtractPlugin, applyCSSLoaders } from './features/css';
import { applyYamlLoader } from './features/yaml';
import { applyTomlLoader } from './features/toml';
import { applyMarkdownLoader } from './features/markdown';
import { applyFallbackLoader } from './features/fallback';
import { applyProgressPlugin } from './features/progress';
import { applyIgnorePlugin } from './features/ignore';
import { applyFileSystemCache } from './features/cache';
import { applyDefinePlugin } from './features/define';

export type ResolveAlias = { [index: string]: string };

class BaseWebpackConfig {
  chain: WebpackChain;

  appContext: IAppContext;

  options: NormalizedConfig;

  appDirectory: string;

  dist: string;

  jsFilename: string;

  jsChunkName: string;

  cssChunkName: string;

  mediaChunkName: string;

  isTsProject: boolean;

  chainUtils: ChainUtils;

  babelPresetAppOptions?: Partial<BabelPrestAppOptions>;

  constructor(appContext: IAppContext, options: NormalizedConfig) {
    this.options = options;
    this.appContext = appContext;
    this.appDirectory = this.appContext.appDirectory;
    this.chain = new WebpackChain();

    const { output = {} } = this.options;
    const { disableAssetsCache } = output;
    const jsPath = (output.jsPath || '').trim();
    const cssPath = (output.cssPath || '').trim();
    const mediaPath = (output.mediaPath || '').trim();

    this.dist = ensureAbsolutePath(this.appDirectory, output.path || '');

    this.jsFilename = removeLeadingSlash(
      `${jsPath}/[name]${
        isProd() && !disableAssetsCache ? '.[contenthash:8]' : ''
      }.js`,
    );

    this.jsChunkName = removeLeadingSlash(
      `${jsPath}/[name]${
        isProd() && !disableAssetsCache ? '.[contenthash:8]' : ''
      }.js`,
    );

    this.cssChunkName = removeLeadingSlash(
      `${cssPath.trim()}/[name]${
        isProd() && !disableAssetsCache ? '.[contenthash:8]' : ''
      }.css`,
    );

    this.mediaChunkName = removeLeadingSlash(
      `${mediaPath}/[name]${disableAssetsCache ? '' : '.[hash:8]'}[ext]`,
    );

    this.isTsProject = isTypescript(this.appDirectory);

    this.chainUtils = {
      chain: this.chain,
      config: this.options,
      loaders: this.chain.module.rule(CHAIN_ID.RULE.LOADERS),
      appContext: this.appContext,
    };
  }

  name() {
    // empty
  }

  target() {
    // empty
  }

  mode() {
    const mode = isProd() ? 'production' : 'development';
    this.chain.mode(mode);
    return mode;
  }

  devtool() {
    const { output } = this.options;
    /* eslint-disable no-nested-ternary */
    this.chain.devtool(
      isProd()
        ? output?.disableSourceMap
          ? false
          : 'source-map'
        : 'cheap-module-source-map',
    );
    /* eslint-enable no-nested-ternary */
  }

  entry() {
    const { entrypoints = [], checkedEntries } = this.appContext;
    const { preEntry } = this.options.source || {};
    const preEntries = preEntry ? ensureArray(preEntry) : [];

    for (const { entryName, entry } of entrypoints) {
      if (checkedEntries && !checkedEntries.includes(entryName)) {
        continue;
      }

      preEntries.forEach(entry => {
        this.chain.entry(entryName).add(entry);
      });

      this.chain.entry(entryName).add(entry);
    }
  }

  output() {
    this.chain.output
      .hashFunction('xxhash64')
      .filename(this.jsFilename)
      .chunkFilename(this.jsChunkName)
      .globalObject('window')
      .path(this.dist)
      .pathinfo(!isProd())
      .devtoolModuleFilenameTemplate(
        // eslint-disable-next-line no-nested-ternary
        isProd()
          ? (info: any) =>
              path
                .relative(
                  this.appContext.srcDirectory,
                  info.absoluteResourcePath,
                )
                .replace(/\\/g, '/')
          : isDev()
          ? (info: any) =>
              path.resolve(info.absoluteResourcePath).replace(/\\/g, '/')
          : undefined,
      )
      .publicPath(this.publicPath());

    this.chain.output.merge({
      assetModuleFilename: this.mediaChunkName,
      environment: {
        arrowFunction: false,
        bigIntLiteral: false,
        const: false,
        destructuring: false,
        dynamicImport: false,
        forOf: false,
        module: false,
      },
    });
  }

  publicPath() {
    const { dev, output } = this.options;

    let publicPath = '/';

    if (isProd()) {
      if (output?.assetPrefix) {
        publicPath = output.assetPrefix;
      }
    } else if (isString(dev?.assetPrefix)) {
      publicPath = dev.assetPrefix;
    } else if (dev?.assetPrefix === true) {
      const ip = this.appContext.ip || 'localhost';
      const port = this.appContext.port || '8080';
      publicPath = `//${ip}:${port}/`;
    }

    if (!publicPath.endsWith('/')) {
      publicPath += '/';
    }

    return publicPath;
  }

  loaders() {
    this.chain.module
      .rule(CHAIN_ID.RULE.MJS)
      .test(/\.m?js/)
      .resolve.set('fullySpecified', false);

    const loaders = this.chain.module.rule(CHAIN_ID.RULE.LOADERS);
    const useTsLoader = Boolean(this.options.output?.enableTsLoader);

    applyBabelLoader({
      ...this.chainUtils,
      useTsLoader,
      babelPresetAppOptions: this.babelPresetAppOptions,
    });

    if (useTsLoader) {
      applyTsLoader(this.chainUtils);
    }

    applyCSSLoaders(this.chainUtils);
    applySvgrLoader({
      ...this.chainUtils,
      mediaChunkName: this.mediaChunkName,
    });
    applyAssetsLoader(this.chainUtils);
    applyYamlLoader(this.chainUtils);
    applyTomlLoader(this.chainUtils);
    applyMarkdownLoader(this.chainUtils);
    applyFallbackLoader(this.chainUtils);

    return loaders;
  }

  plugins() {
    applyProgressPlugin(this.chainUtils);
    applyDefinePlugin(this.chainUtils);
    applyIgnorePlugin(this.chainUtils);

    if (enableCssExtract(this.options)) {
      applyCSSExtractPlugin({
        ...this.chainUtils,
        cssChunkName: this.cssChunkName,
      });
    }

    const { output } = this.options;
    // only enable ts-checker plugin in ts project
    // no need to use ts-checker plugin when using ts-loader
    if (
      this.isTsProject &&
      !output.enableTsLoader &&
      !output.disableTsChecker
    ) {
      applyTsCheckerPlugin(this.chainUtils);
    }
  }

  resolve() {
    // resolve extensions
    const extensions = JS_RESOLVE_EXTENSIONS.filter(
      ext => this.isTsProject || !ext.includes('ts'),
    );

    for (const ext of extensions) {
      this.chain.resolve.extensions.add(ext);
    }

    applyAlias(this.chainUtils);

    //  resolve modules
    this.chain.resolve.modules
      .add('node_modules')
      .add(this.appContext.nodeModulesDirectory);

    // only apply module scope plugin when user config contains moduleScopes
    if (this.options._raw?.source?.moduleScopes) {
      applyModuleScopePlugin(this.chainUtils);
    }

    if (this.isTsProject) {
      applyTsConfigPathsPlugins(this.chainUtils);
    }
  }

  cache() {
    applyFileSystemCache({
      ...this.chainUtils,
      isTsProject: this.isTsProject,
    });
  }

  optimization() {
    const minimize = isProd() && !this.options.output?.disableMinimize;

    this.chain.optimization
      .minimize(minimize)
      .splitChunks({ chunks: 'all' })
      .runtimeChunk({
        name: (entrypoint: any) => `runtime-${entrypoint.name}`,
      });

    if (minimize) {
      applyMinimizer(this.chainUtils);
    }
  }

  stats() {
    this.chain.stats('none');
    this.chain.merge({ infrastructureLogging: getWebpackLogging() });
  }

  config() {
    const chain = this.getChain();
    const chainConfig = chain.toConfig();

    let finalConfig = chainConfig;

    if (this.options.tools?.webpack) {
      let isChainUsed = false;

      const proxiedChain = new Proxy(chain, {
        get(target, property) {
          isChainUsed = true;
          return (target as any)[property];
        },
      });

      const mergedConfig = applyOptionsChain(
        chainConfig,
        this.options.tools?.webpack,
        {
          chain: proxiedChain,
          env: process.env.NODE_ENV!,
          name: chain.get('name'),
          webpack,
          ...getWebpackUtils(chainConfig),
        },
        webpackMerge,
      );

      // Compatible with the legacy `chain` usage, if `chain` is called in `tools.webpack`,
      // using the chained config as finalConfig, otherwise using the merged webpack config.
      if (isChainUsed) {
        finalConfig = chain.toConfig();

        if (isDev()) {
          logger.warn(
            `The ${chalk.cyan('chain')} param of ${chalk.cyan(
              'tools.webpack',
            )} is deprecated, please use ${chalk.cyan(
              'tools.webpackChain',
            )} instead.`,
          );
        }
      } else {
        finalConfig = mergedConfig;
      }
    }

    // TODO remove webpackFinal
    if ((this.options.tools as any)?.webpackFinal) {
      return applyOptionsChain(
        finalConfig,
        (this.options.tools as any)?.webpackFinal,
        {
          name: chain.get('name'),
          webpack,
        },
        webpackMerge,
      );
    }

    return finalConfig;
  }

  applyToolsWebpackChain() {
    if (!this.options.tools) {
      return;
    }

    const { webpackChain } = this.options.tools;
    if (webpackChain) {
      const toArray = <T>(item: T | T[]): T[] =>
        Array.isArray(item) ? item : [item];

      toArray(webpackChain).forEach(item => {
        item(this.chain, {
          env: process.env.NODE_ENV!,
          name: this.chain.get('name'),
          webpack,
          CHAIN_ID,
        });
      });
    }
  }

  getChain() {
    this.chain.context(this.appDirectory);

    this.chain.bail(isProd());
    this.chain.node.set('global', true);

    this.name();
    this.target();
    this.mode();
    this.devtool();
    this.entry();
    this.output();
    this.loaders();
    this.plugins();
    this.resolve();
    this.cache();
    this.optimization();
    this.stats();
    this.applyToolsWebpackChain();

    return this.chain;
  }
}

export { BaseWebpackConfig };
/* eslint-enable max-lines */
