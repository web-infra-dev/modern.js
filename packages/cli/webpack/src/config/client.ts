import {
  isDev,
  isProd,
  CHAIN_ID,
  removeLeadingSlash,
  isFastRefresh,
} from '@modern-js/utils';
import { IAppContext, AppLegacyNormalizedConfig } from '@modern-js/app-tools';
import { HotModuleReplacementPlugin } from 'webpack';
import { BaseWebpackConfig } from './base';
import { addCoreJsEntry } from './features/babel';
import { applyManifestPlugin } from './features/manifest';
import { applyAppIconPlugin } from './features/app-icon';
import { applyCopyPlugin } from './features/copy';
import { applyBundleAnalyzerPlugin } from './features/bundle-analyzer';
import {
  applyReactRefreshPlugin,
  applyReactRefreshBabelPlugin,
} from './features/react-refresh';
import {
  applyNodePolyfillResolve,
  applyNodePolyfillProvidePlugin,
} from './features/node-polyfill';
import { applyEnvVarsDefinePlugin } from './features/define';
import { applyInlineChunkPlugin } from './features/inline-chunk';
import { applyHTMLPlugin } from './features/html';

export class ClientWebpackConfig extends BaseWebpackConfig {
  htmlFilename: (name: string) => string;

  constructor(appContext: IAppContext, options: AppLegacyNormalizedConfig) {
    super(appContext, options);

    this.htmlFilename = (name: string) =>
      removeLeadingSlash(
        `${this.options.output.htmlPath!}/${
          this.options.output.disableHtmlFolder ? name : `${name}/index`
        }.html`,
      );

    this.babelPresetAppOptions = {
      target: 'client',
    };
  }

  name() {
    this.chain.name('client');
  }

  entry() {
    super.entry();
    addCoreJsEntry(this.chainUtils);
  }

  resolve() {
    super.resolve();

    if (!this.options.output.disableNodePolyfill) {
      applyNodePolyfillResolve(this.chainUtils);
    }
  }

  plugins() {
    super.plugins();

    applyEnvVarsDefinePlugin(this.chainUtils);

    const isUsingHMR = isDev() && this.options.tools?.devServer?.hot !== false;
    if (isUsingHMR) {
      this.chain.plugin(CHAIN_ID.PLUGIN.HMR).use(HotModuleReplacementPlugin);

      if (isFastRefresh()) {
        applyReactRefreshPlugin(this.chainUtils);
        applyReactRefreshBabelPlugin(this.chainUtils);
      }
    }

    applyCopyPlugin(this.chainUtils);
    applyHTMLPlugin({
      ...this.chainUtils,
      htmlFilename: this.htmlFilename,
    });
    applyAppIconPlugin(this.chainUtils);
    applyManifestPlugin(this.chainUtils);

    if (isProd()) {
      applyInlineChunkPlugin(this.chainUtils);
    }

    // node polyfill
    if (!this.options.output.disableNodePolyfill) {
      applyNodePolyfillProvidePlugin(this.chainUtils);
    }

    if (this.options.cliOptions?.analyze) {
      applyBundleAnalyzerPlugin({
        ...this.chainUtils,
        reportFilename: 'report.html',
      });
    }
  }
}
