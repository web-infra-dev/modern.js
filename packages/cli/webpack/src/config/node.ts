import type {
  IAppContext,
  ServerUserConfig,
  OutputUserConfig,
  SSGMultiEntryOptions,
  AppLegacyNormalizedConfig,
} from '@modern-js/app-tools';
import { CHAIN_ID, isProd, SERVER_BUNDLE_DIRECTORY } from '@modern-js/utils';
import type WebpackChain from '@modern-js/utils/webpack-chain';
import { BaseWebpackConfig } from './base';
import { applyBundleAnalyzerPlugin } from './features/bundle-analyzer';

export function filterEntriesBySSRConfig(
  chain: WebpackChain,
  serverConfig?: ServerUserConfig,
  outputConfig?: OutputUserConfig,
) {
  const entries = chain.entryPoints.entries();
  // if prod and ssg config is true or function
  if (
    isProd() &&
    (outputConfig?.ssg === true ||
      typeof (outputConfig?.ssg as Array<unknown>)?.[0] === 'function')
  ) {
    return;
  }

  // if single entry has ssg config
  // `ssg: {}` is not allowed if multi entry
  const entryNames = Object.keys(entries);
  if (isProd() && entryNames.length === 1 && outputConfig?.ssg) {
    return;
  }

  // collect all ssg entries
  const ssgEntries: string[] = [];
  if (isProd() && outputConfig?.ssg) {
    const { ssg } = outputConfig;
    entryNames.forEach(name => {
      if ((ssg as SSGMultiEntryOptions)[name]) {
        ssgEntries.push(name);
      }
    });
  }

  const { ssr, ssrByEntries } = serverConfig || {};
  entryNames.forEach(name => {
    if (
      !ssgEntries.includes(name) &&
      ((ssr && ssrByEntries?.[name] === false) ||
        (!ssr && !ssrByEntries?.[name]))
    ) {
      chain.entryPoints.delete(name);
    }
  });
}

class NodeWebpackConfig extends BaseWebpackConfig {
  constructor(appContext: IAppContext, options: AppLegacyNormalizedConfig) {
    super(appContext, options);
    this.babelPresetAppOptions = {
      target: 'server',
      useBuiltIns: false,
    };
  }

  name() {
    this.chain.name('server');
  }

  target() {
    this.chain.target('node');
  }

  entry() {
    super.entry();
    filterEntriesBySSRConfig(
      this.chain,
      this.options.server,
      this.options.output,
    );
  }

  output() {
    super.output();
    this.chain.output
      .libraryTarget('commonjs2')
      .filename(`${SERVER_BUNDLE_DIRECTORY}/[name].js`);

    this.chain.output.delete('chunkFilename');
  }

  // reset ssr bundle optimization
  optimization() {
    super.optimization();
    this.chain.optimization.splitChunks(false as any).runtimeChunk(false);
  }

  loaders() {
    const { USE, ONE_OF } = CHAIN_ID;
    const loaders = super.loaders();

    // css & css modules
    if (loaders.oneOfs.has(ONE_OF.CSS)) {
      loaders.oneOf(ONE_OF.CSS).uses.delete(USE.MINI_CSS_EXTRACT);
      loaders.oneOf(ONE_OF.CSS).uses.delete(USE.STYLE);
    }

    loaders
      .oneOf(ONE_OF.CSS_MODULES)
      .uses.delete(USE.MINI_CSS_EXTRACT)
      .end()
      .uses.delete(USE.STYLE)
      .end()
      .use(USE.CSS)
      .options({
        sourceMap: isProd() && !this.options.output?.disableSourceMap,
        importLoaders: 1,
        modules: {
          localIdentName: this.options.output
            ? this.options.output.cssModuleLocalIdentName!
            : '',
          exportLocalsConvention: 'camelCase',
          exportOnlyLocals: true,
        },
      });

    return loaders;
  }

  plugins() {
    super.plugins();

    // Avoid repeated execution of ts checker
    this.chain.plugins.delete(CHAIN_ID.PLUGIN.TS_CHECKER);

    if (this.options.cliOptions?.analyze) {
      applyBundleAnalyzerPlugin({
        ...this.chainUtils,
        reportFilename: 'report-ssr.html',
      });
    }
  }

  resolve() {
    super.resolve();
    for (const ext of [
      '.node.js',
      '.node.jsx',
      '.node.ts',
      '.node.tsx',
    ].reverse()) {
      this.chain.resolve.extensions.prepend(ext);
    }
    this.chain.resolve.mainFields.clear().add('main');
  }

  config() {
    const config = super.config();

    config.externals = config.externals || [];

    if (!Array.isArray(config.externals)) {
      config.externals = [config.externals].filter(Boolean);
    }

    return config;
  }
}

export { NodeWebpackConfig };
