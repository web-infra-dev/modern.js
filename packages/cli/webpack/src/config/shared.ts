import { ensureArray, WebpackChain } from '@modern-js/utils';
import type {
  IAppContext,
  AppLegacyNormalizedConfig,
} from '@modern-js/app-tools';
import type {
  Configuration,
  RuleSetRule,
  WebpackPluginInstance,
} from 'webpack';
import {
  CSS_REGEX,
  CSS_MODULE_REGEX,
  NODE_MODULES_REGEX,
} from '../utils/constants';

export type ChainUtils = {
  chain: WebpackChain;
  config: AppLegacyNormalizedConfig;
  loaders: WebpackChain.Rule<WebpackChain.Module>;
  appContext: IAppContext;
};

export function getWebpackUtils(config: Configuration) {
  return {
    addRules(rules: RuleSetRule | RuleSetRule[]) {
      const ruleArr = ensureArray(rules);
      config.module?.rules?.unshift(...ruleArr);
    },
    prependPlugins(plugins: WebpackPluginInstance | WebpackPluginInstance[]) {
      const pluginArr = ensureArray(plugins);
      config.plugins?.unshift(...pluginArr);
    },
    appendPlugins(plugins: WebpackPluginInstance | WebpackPluginInstance[]) {
      const pluginArr = ensureArray(plugins);
      config.plugins?.push(...pluginArr);
    },
    removePlugin(pluginName: string) {
      config.plugins = config.plugins?.filter(
        p => p.constructor.name !== pluginName,
      );
    },
  };
}

export const isNodeModulesCss = (path: string) =>
  NODE_MODULES_REGEX.test(path) &&
  CSS_REGEX.test(path) &&
  !CSS_MODULE_REGEX.test(path);
