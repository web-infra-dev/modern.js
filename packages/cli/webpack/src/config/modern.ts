import type {
  IAppContext,
  AppLegacyNormalizedConfig,
} from '@modern-js/app-tools';
import { CHAIN_ID, removeLeadingSlash } from '@modern-js/utils';
import { ClientWebpackConfig } from './client';

class ModernWebpackConfig extends ClientWebpackConfig {
  constructor(appContext: IAppContext, options: AppLegacyNormalizedConfig) {
    super(appContext, options);

    this.htmlFilename = (name: string) =>
      removeLeadingSlash(
        `${this.options.output.htmlPath!}/${
          this.options.output.disableHtmlFolder
            ? `${name}-es6`
            : `${name}/index-es6`
        }.html`,
      );

    this.jsChunkName = this.jsChunkName.replace(/\.js$/, '-es6.js');

    this.jsFilename = this.jsFilename.replace(/\.js$/, '-es6.js');

    this.babelPresetAppOptions = {
      target: 'client',
      useBuiltIns: false,
      useModern: true,
    };
  }

  name() {
    this.chain.name('modern');
  }

  plugins() {
    super.plugins();

    if (this.options.cliOptions?.analyze) {
      this.chain.plugin(CHAIN_ID.PLUGIN.BUNDLE_ANALYZER).tap(() => [
        {
          analyzerMode: 'static',
          openAnalyzer: false,
          reportFilename: 'report-modern.html',
        },
      ]);
    }
  }
}

export { ModernWebpackConfig };
