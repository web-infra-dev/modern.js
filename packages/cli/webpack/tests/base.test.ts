/* eslint-disable-next-line eslint-comments/disable-enable-pair */
/* eslint-disable max-lines */
import path from 'path';
import { NormalizedConfig } from '@modern-js/core';
import type WebpackChain from '@modern-js/utils/webpack-chain';
import { CHAIN_ID } from '@modern-js/utils';
import { BaseWebpackConfig } from '../src/config/base';
import { JS_REGEX, TS_REGEX } from '../src/utils/constants';
import { mergeRegex } from '../src/utils/mergeRegex';
import { userConfig, mockNodeEnv, setPathSerializer } from './util';

describe('base webpack config', () => {
  const fixtures = path.resolve(__dirname, './fixtures');

  const appContext: any = {
    appDirectory: fixtures,
    internalDirectory: '/node_modules/.modern-js',
    srcDirectory: '/src',
    sharedDirectory: './shared',
    entrypoints: [
      {
        entryName: 'page-a',
        entry: path.resolve(fixtures, './demo/src/page-a/index.jsx'),
      },
    ],
    internalDirAlias: '@_modern_js_internal',
    internalSrcAlias: '@_modern_js_src',
  };

  class MyPlugin {
    name: string;

    constructor() {
      this.name = 'MyPlugin';
    }

    // eslint-disable-next-line @typescript-eslint/no-empty-function
    apply() {}
  }

  test(`default webpack config`, () => {
    const config = new BaseWebpackConfig(appContext, {
      ...userConfig,
      source: {
        ...userConfig.source,
        include: ['query-string'],
      },
    } as any).config();

    // todo fix
    // expect(config.output).toEqual(
    //   expect.objectContaining({
    //     filename: 'static/js/[name].js',
    //     chunkFilename: 'static/js/[id].js',
    //     globalObject: 'window',
    //     publicPath: '/',
    //     assetModuleFilename: 'static/media/[name].[hash:8][ext]',
    //   }),
    // );

    expect(config.optimization?.splitChunks).toEqual({ chunks: 'all' });

    expect(config.cache).toHaveProperty('type', 'filesystem');

    expect(config.module.rules).toContainEqual(
      expect.objectContaining({
        oneOf: expect.arrayContaining([
          expect.objectContaining({
            test: mergeRegex(JS_REGEX, TS_REGEX),
          }),
        ]),
      }),
    );
  });

  test(`apply tools.webpackChain`, () => {
    const chainFunction: NormalizedConfig['tools']['webpackChain'] = (
      chain: WebpackChain,
      { env, webpack },
    ) => {
      chain.name('foo');
      expect(env).toEqual('test');
      expect(webpack.DefinePlugin).toBeTruthy();
    };

    const baseConfig = new BaseWebpackConfig(appContext, {
      ...userConfig,
      tools: {
        webpackChain: chainFunction,
      },
    } as any);

    baseConfig.applyToolsWebpackChain();

    expect(baseConfig.config().name).toEqual('foo');
  });

  test(`apply tools.webpack with legacy chain usage`, () => {
    const configFunction: NormalizedConfig['tools']['webpack'] = (
      config,
      { chain },
    ) => {
      chain.name('foo');
    };

    const baseConfig = new BaseWebpackConfig(appContext, {
      ...userConfig,
      tools: {
        webpack: configFunction,
      },
    } as any);

    expect(baseConfig.config().name).toEqual('foo');
  });

  test(`apply tools.webpack and modify the config object`, () => {
    const configFunction: NormalizedConfig['tools']['webpack'] = config => {
      config.name = 'foo';
    };

    const baseConfig = new BaseWebpackConfig(appContext, {
      ...userConfig,
      tools: {
        webpack: configFunction,
      },
    } as any);

    expect(baseConfig.config().name).toEqual('foo');
  });

  test(`apply tools.webpack and return a new config object`, () => {
    const configFunction: NormalizedConfig['tools']['webpack'] = config => {
      config.name = 'foo';
      return {
        name: 'bar',
      };
    };

    const baseConfig = new BaseWebpackConfig(appContext, {
      ...userConfig,
      tools: {
        webpack: configFunction,
      },
    } as any);

    expect(baseConfig.config().name).toEqual('bar');
  });

  test(`apply tools.webpack and using utils.addRules`, () => {
    const newRule = {
      test: /\.foo/,
      loader: 'foo-loader',
    };
    const configFunction: NormalizedConfig['tools']['webpack'] = (
      config,
      utils,
    ) => {
      utils.addRules(newRule);
      utils.addRules([newRule]);
    };

    const baseConfig = new BaseWebpackConfig(appContext, {
      ...userConfig,
      tools: {
        webpack: configFunction,
      },
    } as any);

    expect(baseConfig.config().module.rules[0]).toEqual(newRule);
    expect(baseConfig.config().module.rules[1]).toEqual(newRule);
  });

  test(`apply tools.webpack and using utils.prependPlugins`, () => {
    const configFunction: NormalizedConfig['tools']['webpack'] = (
      config,
      utils,
    ) => {
      utils.prependPlugins(new MyPlugin());
      utils.prependPlugins([new MyPlugin()]);
    };

    const baseConfig = new BaseWebpackConfig(appContext, {
      ...userConfig,
      tools: {
        webpack: configFunction,
      },
    } as any);

    expect(baseConfig.config().plugins[0].constructor.name).toEqual('MyPlugin');
    expect(baseConfig.config().plugins[1].constructor.name).toEqual('MyPlugin');
  });

  test(`apply tools.webpack and using utils.appendPlugins`, () => {
    const configFunction: NormalizedConfig['tools']['webpack'] = (
      config,
      utils,
    ) => {
      utils.appendPlugins(new MyPlugin());
      utils.appendPlugins([new MyPlugin()]);
    };

    const baseConfig = new BaseWebpackConfig(appContext, {
      ...userConfig,
      tools: {
        webpack: configFunction,
      },
    } as any);

    const { plugins } = baseConfig.config();
    expect(plugins[plugins.length - 2].constructor.name).toEqual('MyPlugin');
    expect(plugins[plugins.length - 1].constructor.name).toEqual('MyPlugin');
  });

  test(`apply tools.webpack and using utils.removePlugin`, () => {
    const configFunction: NormalizedConfig['tools']['webpack'] = (
      config,
      utils,
    ) => {
      utils.appendPlugins([new MyPlugin()]);
      utils.removePlugin('MyPlugin');
    };

    const baseConfig = new BaseWebpackConfig(appContext, {
      ...userConfig,
      tools: {
        webpack: configFunction,
      },
    } as any);

    const { plugins } = baseConfig.config();
    expect(
      plugins[plugins.length - 1].constructor.name === 'MyPlugin',
    ).toBeFalsy();
  });

  test(`should using output.assetPrefix as publicPath in dev`, () => {
    const restore = mockNodeEnv('development');

    const baseConfig = new BaseWebpackConfig(appContext, {
      ...userConfig,
      dev: {
        ...userConfig.dev,
        assetPrefix: 'https://dev/',
      },
      output: {
        ...userConfig.output,
        assetPrefix: 'https://prod/',
      },
    } as any);

    const { output } = baseConfig.config();
    expect(output.publicPath).toEqual('https://dev/');

    restore();
  });

  test(`should using dev.assetPrefix as publicPath in prod`, () => {
    const restore = mockNodeEnv('production');

    const baseConfig = new BaseWebpackConfig(appContext, {
      ...userConfig,
      dev: {
        ...userConfig.dev,
        assetPrefix: 'https://dev/',
      },
      output: {
        ...userConfig.output,
        assetPrefix: 'https://prod/',
      },
    } as any);

    const { output } = baseConfig.config();
    expect(output.publicPath).toEqual('https://prod/');

    restore();
  });

  test(`apply tools.tsLoader and using utils.addIncludes/addExcludes`, () => {
    const configFunction: NormalizedConfig['tools']['tsLoader'] = (
      config,
      utils,
    ) => {
      utils.addIncludes(/include-1/);
      utils.addIncludes([/include-2/, /include-3/]);
      utils.addExcludes(/exclude-1/);
      utils.addExcludes([/exclude-2/, /exclude-3/]);
    };

    const baseConfig = new BaseWebpackConfig(appContext, {
      ...userConfig,
      output: {
        ...userConfig.output,
        enableTsLoader: true,
        path: `${__dirname}/dist`,
      },
      tools: {
        tsLoader: configFunction,
      },
    } as any);

    const rule = baseConfig
      .getChain()
      .module.rule(CHAIN_ID.RULE.LOADERS)
      .oneOf(CHAIN_ID.ONE_OF.TS);

    setPathSerializer();
    expect(rule.include.values()).toMatchSnapshot();
    expect(rule.exclude.values()).toMatchSnapshot();
  });

  test(`apply tools.babel and using utils.addIncludes/addExcludes`, () => {
    const configFunction: NormalizedConfig['tools']['babel'] = (
      config,
      utils,
    ) => {
      utils.addIncludes(/include-1/);
      utils.addIncludes([/include-2/, /include-3/]);
      utils.addExcludes(/exclude-1/);
      utils.addExcludes([/exclude-2/, /exclude-3/]);
    };

    const baseConfig = new BaseWebpackConfig(appContext, {
      ...userConfig,
      output: {
        ...userConfig.output,
        path: `${__dirname}/dist`,
      },
      tools: {
        babel: configFunction,
      },
    } as any);

    const rule = baseConfig
      .getChain()
      .module.rule(CHAIN_ID.RULE.LOADERS)
      .oneOf(CHAIN_ID.ONE_OF.JS);

    setPathSerializer();
    expect(rule.include.values()).toMatchSnapshot();
    expect(rule.exclude.values()).toMatchSnapshot();
  });

  test(`should apply module scope plugin when user config contains moduleScopes`, () => {
    const baseConfig = new BaseWebpackConfig(appContext, {
      ...userConfig,
      output: {
        ...userConfig.output,
        path: `${__dirname}/dist`,
      },
      _raw: {
        source: {
          moduleScopes: ['./foo'],
        },
      },
    } as any);

    const moduleScopePlugin = baseConfig
      .getChain()
      .resolve.plugins.get(CHAIN_ID.RESOLVE_PLUGIN.MODULE_SCOPE);

    expect(moduleScopePlugin).toBeTruthy();
  });

  test(`should not apply module scope plugin when user config not contains moduleScopes`, () => {
    const baseConfig = new BaseWebpackConfig(appContext, {
      ...userConfig,
      output: {
        ...userConfig.output,
        path: `${__dirname}/dist`,
      },
    } as any);

    const moduleScopePlugin = baseConfig
      .getChain()
      .resolve.plugins.get(CHAIN_ID.RESOLVE_PLUGIN.MODULE_SCOPE);

    expect(moduleScopePlugin).toBeFalsy();
  });

  test(`should prepend entries correctly when using source.preEntry config`, () => {
    const config = new BaseWebpackConfig(appContext, {
      ...userConfig,
      source: {
        ...userConfig.source,
        preEntry: ['./src/foo.ts'],
      },
    } as any).config();

    expect(config.entry['page-a']?.[0]).toEqual('./src/foo.ts');
  });
});
