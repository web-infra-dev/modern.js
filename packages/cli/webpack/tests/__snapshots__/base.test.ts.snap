// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base webpack config apply tools.babel and using utils.addIncludes/addExcludes 1`] = `
Array [
  Object {
    "and": Array [
      /cli/webpack/tests/fixtures,
      Object {
        "not": /node_modules/,
      },
    ],
  },
  /.modern-js,
  /cli/webpack/src/runtime/core-js-entry.js,
  /include-1/,
  /include-2/,
  /include-3/,
]
`;

exports[`base webpack config apply tools.babel and using utils.addIncludes/addExcludes 2`] = `
Array [
  /exclude-1/,
  /exclude-2/,
  /exclude-3/,
]
`;

exports[`base webpack config apply tools.tsLoader and using utils.addIncludes/addExcludes 1`] = `
Array [
  Object {
    "and": Array [
      /cli/webpack/tests/fixtures,
      Object {
        "not": /node_modules/,
      },
    ],
  },
  /.modern-js,
  /cli/webpack/src/runtime/core-js-entry.js,
  /include-1/,
  /include-2/,
  /include-3/,
]
`;

exports[`base webpack config apply tools.tsLoader and using utils.addIncludes/addExcludes 2`] = `
Array [
  /exclude-1/,
  /exclude-2/,
  /exclude-3/,
]
`;
